engine/src/eventqueue.cpp

```diff
@@ -490,32 +491,37 @@ static void MCEventQueueDispatchEvent(MCEvent *p_event)
 	case kMCEventTypeKeyPress:
 		{
 			MCStackHandle t_stack = t_event->key.stack;
-            
-            MCObject *t_target = t_menu != nil ? t_menu : t_stack;
+
+			MCObject *t_target = t_menu != nil ? t_menu : t_stack;
 
 			MCmodifierstate = t_event -> key . press . modifiers;
 
-			// If 'char_code' is 0, then this key press has not generated a
-			// character.
+			MCAutoStringRef t_char;
+
 			if (t_event -> key . press . char_code == 0)
 			{
-				t_target -> kdown(kMCEmptyString, t_event -> key . press . key_code);
-				t_target -> kup(kMCEmptyString, t_event -> key . press . key_code);
-				break;
+				// If 'char_code' is 0, then this key press has not generated a
+				// character.
+				t_char = kMCEmptyString;
 			}
+			else
+			{
+				// Otherwise 'char_code' is the unicode codepoint, so first map to
+				// UTF-16 codeunits
+				unichar_t t_unichars[2];
+				uindex_t t_length = 1;
+				if (MCUnicodeCodepointToSurrogates(t_event->key.press.char_code, t_unichars[0], t_unichars[1]))
+					t_length = 2;
+
+				// Now the string is created with the appropriate unicode-capable function
+				MCStringCreateWithChars(t_unichars, t_length, &t_char);
+			}
+
+			if (t_event->key.press.state == kMCEventKeyStateDown || t_event->key.press.state == kMCEventKeyStatePressed)
+				t_target -> kdown(*t_char, t_event -> key . press . key_code);
 
-			// Otherwise 'char_code' is the unicode codepoint, so first map to
-			// UTF-16 codeunits
-			unichar_t t_unichars[2];
-            uindex_t t_length = 1;
-            if (MCUnicodeCodepointToSurrogates(t_event->key.press.char_code, t_unichars[0], t_unichars[1]))
-                t_length = 2;
-
-			// Now the string is created with the appropriate unicode-capable function
-			MCAutoStringRef t_buffer;
-            MCStringCreateWithChars(t_unichars, t_length, &t_buffer);
-			t_target -> kdown(*t_buffer, t_event -> key . press . key_code);
-			t_target -> kup(*t_buffer, t_event -> key . press . key_code);
+			if (t_event->key.press.state == kMCEventKeyStateUp || t_event->key.press.state == kMCEventKeyStatePressed)
```

Comment:
I think this is a departure from the current behavior. I'm pretty sure on Desktop (and also mobile) we only really do anything with Down - which causes a down and then an up to be synthesized. So a down/up pair is a key press (which may repeat at repeat rate!) *not* an indication of the actual key down or up state (as in the key on the keyboard).

Definitely worth comparing and contrasting with desktop!

Reply:
@runrevmark I'm confused by this comment - are you saying that on all platforms key up messages are ignored? From looking at the desktop implementations that's certainly not the case - Mac, Win, and Linux all process key up events and forward them to the active stack. I know that on mobile platforms we synthesize key down+up events from a single (virtual) keypress but that behaviour has been retained - the kEventKeyStatePressed state results in down + up. The only change is in the HTML5 engine where keydown & keyup events are processed instead of the deprecated keypress event, but this should result in the same, intended behaviour.

Reply:
See also Line 49 of [em-event.js](https://github.com/livecode/livecode/blob/develop/engine/src/em-event.js)